demonstrations/spsa.py

```diff
@@ -0,0 +1,734 @@
+stochasticity of the technique comes from the fact that for each iteration step
+:math:`k` the components of the :math:`\Delta_{k}` perturbation vector are
+randomly generated using a zero-mean distribution. In most cases, the Bernoulli
+distribution is used, meaning each parameter is simultaneously perturbed by
+either :math:`\pm c_k`.
+
+It is this perturbation that makes SPSA robust to noise --- since every
+parameter is already being shifted, additional shifts due to noise are less
+likely to hinder the optimization process. In a sense, noise gets "absorbed"
+into the already-stochastic process. This is highlighted in the figure below,
+which portrays an example of the type of path SPSA takes through the space of
+the function, compared to a standard gradient-based optimizer.
+
+.. figure:: ../demonstrations/spsa/spsa_opt.png
+   :align: center
+   :width: 60%
+
+   ..
+
+   A schematic of the search paths used by gradient descent with
+   parameter-shift and SPSA in a low-noise setting.
+   Image source: [#spall_overview]_
+
+Now that we have explored how SPSA works, let's see how it performs in practice!
+
+Optimization on a sampling device
+---------------------------------
+
+.. important::
+
+    To run this demo locally, you'll need to install the `noisyopt
+    <https://github.com/andim/noisyopt>`_ library. This library contains a
+    straightforward implementation of SPSA that can be used in the same way as the
+    optimizers available in `SciPy's minimize method
+    <https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html>`_.
+
+First, let's consider a simple quantum circuit on a sampling device. For this,
+we'll be using a device from the `PennyLane-Qiskit plugin
+<https://pennylaneqiskit.readthedocs.io/en/latest/>`_ that samples quantum
+circuits to get measurement outcomes and later post-processes these outcomes to
+compute statistics like expectation values.
+
+.. note::
+
+    Just with other PennyLane devices, the number of samples taken for a device
+    execution can be specified using the ``shots`` keyword argument of the
+    device.
+
+Once we have a device selected, we just need a couple of other ingredients for
+the pieces of an example optimization to come together:
+
+* a circuit ansatz: :func:`~.pennylane.templates.layers.StronglyEntanglingLayers`,
+* initial parameters: conveniently generated using :func:`~.pennylane.init.strong_ent_layers_normal`,
+* an observable: :math:`\bigotimes_{i=0}^{N-1}\sigma_z^i`, where :math:`N` stands
+  for the number of qubits.
+
+"""
+import pennylane as qml
+import numpy as np
+
+num_wires = 4
+num_layers = 5
+
+dev_sampler_spsa = qml.device("qiskit.aer", wires=num_wires, shots=1000)
+
+##############################################################################
+# We seed so that we can simulate the same circuit every time.
+np.random.seed(50)
+
+all_pauliz_tensor_prod = qml.operation.Tensor(
+    *[qml.PauliZ(i) for i in range(num_wires)]
+)
```

Comment:
```suggestion
all_pauliz_tensor_prod = qml.operation.Tensor(*[qml.PauliZ(i) for i in range(num_wires)])
```

This got expanded when I ran the code through Black (with a conservative number of lines), so we can either keep as is or put back to the original - it doesn't cause a horizontal scrollbar.

Reply:
Thank you for checking that! The one-liner seems good